src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1252,1807 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+		D_GOTO(out, rc = -DER_NOMEM);
+	hdf5.oid = &(hdf5.oid_data);
+	hdf5.dk = &(hdf5.dkey_data);
+	hdf5.ak = &(hdf5.akey_data);
+	hdf5.oid_buf_size = sizeof(struct oid);
+	hdf5.dkey_buf_size = sizeof(struct dkey);
+	hdf5.akey_buf_size = sizeof(struct akey);
+
+	/* create snapshot, open oit, then starting iterating over oids */
+	rc = daos_cont_create_snap_opt(coh, &epoch, NULL,
+				       DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to create snapshot: "DF_RC, DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+	rc = daos_oit_open(coh, epoch, &toh, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to open object iterator: "DF_RC, DP_RC(rc));
+		D_GOTO(out_snap, rc);
+	}
+	memset(&anchor, 0, sizeof(anchor));
+	while (!daos_anchor_is_eof(&anchor)) {
+		oids_nr = OID_ARR_SIZE;
+		rc = daos_oit_list(toh, oids, &oids_nr, &anchor, NULL);
+		if (rc != 0) {
+			D_ERROR("Failed to list objects: "DF_RC, DP_RC(rc));
+			D_GOTO(out_oit, rc);
+		}
+
+		hdf5.oid_old_buf_size = hdf5.oid_buf_size;
+		hdf5.oid_buf_size = (oids_nr + *total_oids) * sizeof(struct oid);
+		D_REALLOC(oid_tmp, hdf5.oid_data, hdf5.oid_old_buf_size, hdf5.oid_buf_size);
+		if (oid_tmp == NULL)
+			D_GOTO(out_oit, rc = -DER_NOMEM);
+		hdf5.oid_data = oid_tmp;
+
+		/* list object ID's */
+		for (i = 0; i < oids_nr; i++) {
+			/* save oid hi and low in oid buffer */
+			(*hdf5.oid)[*total_oids].oid_hi = oids[i].hi;
+			(*hdf5.oid)[*total_oids].oid_low = oids[i].lo;
+
+			is_kv = obj_is_kv(oids[i]);
+			if (is_kv) {
+				rc = daos_kv_open(coh, oids[i], DAOS_OO_RW, &oh, NULL);
+				if (rc != 0) {
+					D_ERROR("Failed to open kv object: "DF_RC, DP_RC(rc));
+					D_GOTO(out_oit, rc);
+				}
+
+				/* serialize dkeys */
+				rc = serialize_dkeys(&hdf5, oids[i], coh, total_oids, &dkey_index,
+						     &akey_index, &oh, is_kv, total_dkeys,
+						     total_akeys, bytes_read);
+				if (rc != 0) {
+					D_ERROR("Failed to serialize dkeys: "DF_RC, DP_RC(rc));
+					D_GOTO(err_kv_obj, rc);
+				}
+
+				rc = daos_kv_close(oh, NULL);
+				if (rc != 0) {
+					D_ERROR("Failed to close kv object: "DF_RC, DP_RC(rc));
+					D_GOTO(out_oit, rc);
+				}
+			} else {
+				rc = daos_obj_open(coh, oids[i], 0, &oh, NULL);
+				if (rc != 0) {
+					D_ERROR("Failed to open object: "DF_RC, DP_RC(rc));
+					D_GOTO(out_oit, rc);
+				}
+
+				/* serialize dkeys */
+				rc = serialize_dkeys(&hdf5, oids[i], coh, total_oids, &dkey_index,
+						     &akey_index, &oh, is_kv, total_dkeys,
+						     total_akeys, bytes_read);
+
+				if (rc != 0) {
+					D_ERROR("Failed to serialize dkeys: "DF_RC, DP_RC(rc));
+					D_GOTO(err_obj, rc);
+				}
+
+				rc = daos_obj_close(oh, NULL);
+				if (rc != 0) {
+					D_ERROR("Failed to close object: "DF_RC, DP_RC(rc));
+					D_GOTO(out_oit, rc);
+				}
+			}
+			(*total_oids)++;
+		}
+	}
+
+	rc = write_oid_dset(&hdf5, *total_oids);
+	if (rc != 0)
+		D_GOTO(out_oit, rc);
+
+	/* create dkey dataset and write dkeys */
+	rc = write_dkey_dset(&hdf5, *total_dkeys);
```

Comment:
again i don't think it's feasible to pull the entire dkey and akey dataset into memory then write it to the HDF5 file. this would cause OOM on most real datasets.